File: android/drengr/src/test/kotlin/dev/drengr/sdk/ClassifyTest.kt
package dev.drengr.sdk

import org.json.JSONObject
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test

/** Fail-closed parity suite for the seal-by-default classifier (mirrors classify.ts). */
class ClassifyTest {
    private val email = "john.doe@example.com"
    private val phone = "555-123-4567"
    private val ssn = "123-45-6789"
    private val ip = "192.168.1.42"
    private val uuid = "550e8400-e29b-41d4-a716-446655440000"
    private val name = "Jonathan Q. Public"
    private val address = "221B Baker Street"
    private val leaks = listOf(email, phone, ssn, ip, uuid, name, address)

    @Test fun bodySealsAllPiiKeepsBusiness() {
        val body = """{
            "email":"$email","phone":"$phone","ssn":"$ssn","clientIp":"$ip",
            "deviceId":"$uuid","firstName":"$name","address":"$address",
            "amount":25,"currency":"USD","status":"DECLINED","ok":true,"count":3
        }"""
        val proj = Classify.classifyBody(body).projection!!
        for (leak in leaks) assertFalse("leaked: $leak", proj.contains(leak))
        val o = JSONObject(proj)
        assertEquals(25, o.getInt("amount"))
        assertEquals("USD", o.getString("currency"))
        assertEquals("DECLINED", o.getString("status"))
        assertTrue(o.getBoolean("ok"))
        assertEquals(3, o.getInt("count"))
    }

    @Test fun valueDetectedPiiUnderBenignName() {
        // benign field names, PII by value → still sealed
        val proj = Classify.classifyBody("""{"note":"$email","ref":"$ip","tag":"$uuid"}""").projection!!
        for (leak in listOf(email, ip, uuid)) assertFalse(leak, proj.contains(leak))
    }

    @Test fun typedPlaceholderPreservesType() {
        val o = JSONObject(Classify.classifyBody("""{"lat":37.77,"lng":-122.41,"count":5}""").projection!!)
        // lat/lng are PII names with numeric values → numeric placeholder, not a string
        assertEquals(0, o.getInt("lat"))
        assertEquals(0, o.getInt("lng"))
        assertEquals(5, o.getInt("count"))
    }

    @Test fun freeTextStringSealsByDefault() {
        // unknown string name, no detectable pattern → seal-by-default
        val o = JSONObject(Classify.classifyBody("""{"comment":"hello there friend","status":"OK"}""").projection!!)
        assertEquals("[PII]", o.getString("comment"))
        assertEquals("OK", o.getString("status"))
    }

    @Test fun credentialsDropped() {
        val proj = Classify.classifyBody("""{"password":"hunter2secret","card":"4111 1111 1111 1111","amount":25}""").projection!!
        assertFalse(proj.contains("hunter2secret"))
        assertFalse(proj.contains("4111"))
        assertTrue(proj.contains("REDACTED-SECRET"))
        assertEquals(25, JSONObject(proj).getInt("amount"))
    }

    @Test fun headerPiiMaskedAndValueScrubbed() {
        val out = Redact.redactHeaders(
            mapOf(
                "X-User-Email" to email,
                "X-Client-Ip" to ip,
                "X-Device-Id" to uuid,
                "Accept" to "application/json",
            ),
            emptySet(),
        )
        assertEquals("[REDACTED]", out["X-User-Email"]) // PII-denoting name → full mask
        for (leak in listOf(email, ip, uuid)) assertFalse(leak, out.values.any { it.contains(leak) })
        assertEquals("application/json", out["Accept"])
    }

    @Test fun urlPiiScrubbed() {
        val r = Redact.redactUrl("https://api.x.com/u?email=$email&ip=$ip&device=$uuid&amount=25")
        for (leak in listOf(email, ip, uuid)) assertFalse(leak, r.contains(leak))
        assertTrue(r.contains("amount=25"))
    }

    @Test fun failOpen() {
        assertNull(Classify.classifyBody("").projection)
        assertNull(Classify.classifyBody(null).projection)
        assertNull(Classify.classifyBody("not json not form").projection)
        assertNull(Classify.classifyBody("{unterminated").projection)
    }
}

File: flutter/test/classify_test.dart
import 'dart:convert';

import 'package:drengr_flutter_sdk/src/classify.dart';
import 'package:drengr_flutter_sdk/src/redact.dart';
import 'package:test/test.dart';

void main() {
  group('classify (seal-by-default)', () {
    test('fail-closed: planted PII never reaches the projection', () {
      final body = jsonEncode({
        'email': 'alice@example.com',
        'phone': '+1 415-555-1212',
        'ssn': '123-45-6789',
        'ip': '192.168.1.42',
        'deviceId': 'AB12CD34-1234-5678-9ABC-DEF012345678',
        'fullName': 'Alice Robinson',
        'address': '221B Baker Street',
        'password': 'hunter2secret',
        'note': 'free text that could hide alice@example.com',
        'amount': 4200,
        'currency': 'USD',
        'success': true,
        'status': 'ok',
      });
      final c = classifyBody(body);
      final proj = c.projection!;
      for (final leak in <String>[
        'alice@example.com',
        '415-555-1212',
        '123-45-6789',
        '192.168.1.42',
        'AB12CD34-1234-5678-9ABC-DEF012345678',
        'Alice Robinson',
        '221B Baker Street',
        'hunter2secret',
        'free text',
      ]) {
        expect(proj, isNot(contains(leak)), reason: 'leaked $leak');
      }
      final m = jsonDecode(proj) as Map<String, dynamic>;
      // Business signal survives plaintext.
      expect(m['amount'], 4200);
      expect(m['currency'], 'USD');
      expect(m['success'], true);
      expect(m['status'], 'ok');
      // Typed placeholder keeps leaf type stable (string field -> string).
      expect(m['email'], isA<String>());
      // Credential is dropped, not sealed.
      expect(c.piiPaths, isNot(contains('password')));
      // PII paths are collected (for the encrypt layer) but not shipped here.
      expect(c.piiPaths, contains('email'));
    });

    test('typed placeholder preserves number/bool type under a PII name', () {
      final c = classifyBody(jsonEncode({'lat': 37.77, 'geo': true}));
      final m = jsonDecode(c.projection!) as Map<String, dynamic>;
      expect(m['lat'], 0); // number -> 0
      expect(m['geo'], false); // bool -> false
    });

    test('credential value under a benign name is dropped, never sealed', () {
      final c = classifyBody(
          jsonEncode({'data': 'sk_live_abcdef0123456789ABCDEF', 'orderId': 7}));
      final proj = c.projection!;
      expect(proj, contains('[REDACTED-SECRET]'));
      expect(proj, isNot(contains('sk_live_abcdef')));
      expect(c.piiPaths, isEmpty);
      final m = jsonDecode(proj) as Map<String, dynamic>;
      expect(m['orderId'], 7);
    });

    test('form-encoded seals PII by default, keeps allowlisted', () {
      final c = classifyBody('email=a%40b.com&status=ok&amount=25');
      final proj = c.projection!;
      expect(proj, isNot(contains('a@b.com')));
      expect(proj, isNot(contains('a%40b.com')));
      final m = jsonDecode(proj) as Map<String, dynamic>;
      expect(m['status'], 'ok');
    });

    test('empty / non-structured bodies yield an empty split', () {
      expect(classifyBody(null).projection, isNull);
      expect(classifyBody('').projection, isNull);
      expect(classifyBody('plain hello world').projection, isNull);
    });
  });

  group('redaction hardening (IPv4 + UUID + header names)', () {
    test('scrubValues masks IPv4 and UUID anywhere in text', () {
      expect(scrubValues('client 192.168.1.42 connected'),
          isNot(contains('192.168.1.42')));
      expect(scrubValues('id=AB12CD34-1234-5678-9ABC-DEF012345678'),
          isNot(contains('AB12CD34-1234-5678-9ABC-DEF012345678')));
    });

    test('redactHeaders masks PII-denoting names AND scrubs PII values', () {
      final r = redactHeaders({
        'X-User-Phone': 'bob', // PII by name -> full mask
        'X-Client-IP': '192.168.1.42', // PII by value under benign name
        'X-Device-Id': 'AB12CD34-1234-5678-9ABC-DEF012345678',
        'Accept': 'application/json',
      }, const {});
      expect(r['X-User-Phone'], redactMask);
      expect(r['X-Client-IP'], isNot(contains('192.168.1.42')));
      expect(
          r['X-Device-Id'], isNot(contains('AB12CD34-1234-5678-9ABC-DEF012345678')));
      expect(r['Accept'], 'application/json');
    });
  });
}
